src/cli/commands/add.js

```diff
@@ -51,16 +51,46 @@ export class Add extends Install {
         optional: false,
       });
     }
+    return requestsWithArgs;
+  }
 
-    return Promise.resolve({
-      patterns: patterns.concat(this.args),
-      requests: requestsWithArgs,
-    });
+  preparePatterns(
+    patterns: Array<string>,
+  ): Array<string> {
+    const {exact, tilde} = this.flags;
+    const preparedPatterns = patterns.slice();
+    for (const pattern of this.resolver.dedupePatterns(this.args)) {
+      const pkg = this.resolver.getResolvedPattern(pattern);
+      invariant(pkg, `missing package ${pattern}`);
+
+      const parts = PackageRequest.normalizePattern(pattern);
```

Comment:
A lot of this code is copied from this same file, can we colocate this into a method that's shared?